### DISCUSSION TITLE
Is it possible to use the input to control the uiMidiKeyboard? Or to rechannel the MIDI input notes?

### DISCUSSION THREAD

**Discussion by dervur** [RomanKubiak/ctrlr#669] [Q&A] [answered]
Hello there,

I'm trying to find a solution to re-channelize the midi input of a keyboard controller set as input via Ctrlr. I couldn't find any solution in the panel settings, the only way I can send the notes to the output is by activating the input to output option ( the one with (channelize)" does not seem to do anything.
Therefore I was thinking if it is possible to use the keyboard input to simply control the uiMidiKeyboard. In this way, whatever channel I set in the output the uiMidiKeyboard will follow and send there. I don't know if this is possible without using LUA or if I need LUA to snoop the input note and send it to uiMidiKeyboard. 

This is the only way I'm thinking it would allow a re-channelize of the input. If there are any other ways I'm missing please let me know. I tried to look in the original forum to no avail and another post here on github was mentioning a setting Ctrlr Controller to Output that would allow this, but I don't see it.

**Comment by damiensellier** [comment]
I did it on a panel to convert midi input messages to Mackie MCU. It requires a transformer script in LUA on  the  "when the panel receives a midi message" callback in the global properties.
Message comes in, you translate it with the required modifications and send the new midi message from the script.
You can do any midi transformation like this. Channel change, value change, cc number change etc

**Reply by dervur** [reply]
Mmmh I was thinking to do that as Plan B. Do all the functions for the MIDI works? Since some MIDI properties for modulators make Ctrlr crash (https://github.com/RomanKubiak/ctrlr/discussions/667) I was avoiding that as first solution on purpose. 
is there a function to get the channel of the MIDI input? Or should I take the overall MIDI message, change the specific hex of the channel and then send it to the output?

**Reply by damiensellier** [reply]
You need to store the midimessage as memory block and work with arithmetic functions to transform the different bytes as required. This is the first thing that comes to mind about it. Then use the usual sendmessage now function with the message as memory block.

**Reply by dervur** [reply]
Thank you very much. Wouldn't it work like in the code below? So just using the functions to manipulate the CtrlrMidiMessage 

```
-- Called when a MIDI input arrives to the panel
rechannelMidiNotes = function(--[[ CtrlrMidiMessage --]] midi)

local channelOut = panel:getProperty("panelMidiOutputChannelDevice")
midi:setChannel(channelOut)
panel:sendMidiMessageNow(midi)

end
```

I guess I can then add an if statment to see if the MIDI message is a note and not something else I than should use midi:getMidiMessageType().

I don't have a controller now that I'm writing but I'll give it a go, unless you already know it will not work.

**Reply by damiensellier** [reply]
Sorry but I have no idea.

**Reply by dervur** [reply]
I just tested it and it works by converting channelOut to number (just used the `tonumber()` function). Kinda easier than I thought.
However I saw that sendMidiMessageNow also sends in the Midi input is there a way to only send at the output? I mean it doesn't really matter in my scenario but I think it would look cleaner if the message is sent to only the output.

Here I was playing with the controller on Ch.1 and the Channel Output was set to Ch.2
![image](https://github.com/user-attachments/assets/4ab9faad-393a-4b44-a132-540eb5e036ec)
The output shows only one note on and one off, but the input first shows the Ch.1 note and then sends the Ch.2 note too. I'm just thinking about scenarios where the note sent is continuous or more complex (like with aftertouch) this could maybe create issues (or maybe not), but I think it would look cleaner if at the input there is only the Ch.1 note and at the output there is a Ch.2 note. If you don't know it's fine I was just wondering.

**Reply by dervur** [reply]
Ok no, I noticed that the note is doubled even without using the function, so I think the issue is MIDI Tools or loopMIDI, I'll try with a proper MIDI controller when I'm at home. But I saw that even with the MIDI thru device to device on, in the input the message appears twice, in the output it appears only once. I'll post below the full code and mark it as answer.

EDIT: I am an idiot, I had the MIDI output for one panel set as the MIDI input for another panel in the session of CtrlrX, thence the double MIDI message 🤦‍♂️

**Comment by dervur** [comment, marked as answer]
Ok I solved it this way. I have a button to control if the input should be re-channeled, with a method when it changes the status that activates the midi thru input to output if the button is off and viceversa it deactivates it if the button is on.

```
rechannelMidiButton = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

    local rechannelInput = mod:getValue()

    if rechannelInput == 1 then
        panel:setProperty("panelMidiThruD2D",0,false)
    else
        panel:setProperty("panelMidiThruD2D",1,false)
    end

end
```

And then I have the following method whenever I receive a MIDI input on the panel
```
rechannelMidiNotes = function(--[[ CtrlrMidiMessage --]] midi)
    --Get the value of the rechannel button
    local rechannelInput = panel:getModulatorByName("rechannelButton"):getValue()

    if rechannelInput == 1 then
        -- Get the set channel output and convert it from string to number
        local channelOut = tonumber(panel:getProperty("panelMidiOutputChannelDevice"))
        -- Change the channel of the incoming MIDI message to the channel set at the output
        midi:setChannel(channelOut)
        -- Send MIDI message
        panel:sendMidiMessageNow(midi)
    end
end
```

I mean I guess you could do without the MIDI thru, by adding an else in `rechannelMidiNotes`, but that's how I did it and it works...
Also I think I should include one statement for the MIDI thru as well as VST, but it just needs expanded from this code, I was only testing it from the exe.

EDIT: in reality I'm rechanneling everything from the MIDI input not the notes only, I think it should be expanded to treat only the notes, but for my use it is fine like this.
